common/lib/capa/capa/responsetypes.py

```diff
@@ -720,14 +726,103 @@ def assign_choice_names(self):
 
     def get_score(self, student_answers):
 
+        """
+        Answer lists:
+          all_choices: the full list of checkboxes
+          student_answer: what the student actually chose (note no "s")
+          student_non_answer: what they didn't choose
+          self.correct_choices: boxes that should be checked
+          self.incorrect_choices: boxes that should NOT be checked
+        """
+        
+        # pipe for union of sets
+        all_choices = self.correct_choices | self.incorrect_choices   
+        
         student_answer = student_answers.get(self.answer_id, [])
 
         if not isinstance(student_answer, list):
             student_answer = [student_answer]
 
         no_empty_answer = student_answer != []
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but it throws an error if all the checks are blank.
+        
         student_answer = set(student_answer)
 
+        student_non_answers = all_choices - student_answer
+        
+        if not no_empty_answer:
+            print "Empty Answer"
+            return CorrectMap(self.answer_id, 'incorrect')
+        
+        
+        """
+        This below checks to see whether we're using an alternate grading scheme.
+          Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+          Set partial_credit="EDC" to count each choice for equal points (Every Decision Counts).
+          Set partial_credit="halves" to take half credit off for each error.
+        """
+        
+        tree = self.xml
+        problem_xml = tree.xpath('.')
+
+        # Partial credit type - only one type at a time right now.
+        credit_type = problem_xml[0].get('partial_credit', default=False)
+
+        try:
+            credit_type = str(credit_type).lower().strip()
+        except ValueError:
+            _ = self.capa_system.i18n.ugettext
```

Comment:
I am a bit confused as to why there would be a substantial performance difference, but that may be my own misunderstanding of the code and how XML gets passed around. 

It does feel like this gets repeated a lot. Under DRY, I'd personally move the first two lines into either LoncapaReponse (self.xml_tree = tree.xpath(.) or similar) or redefining the convention to have LoncapaProblem pass in an xml tree instead of an XML node. I kind of prefer the former, since a response type might need access to things outside of its tree. 

Overall, it's a question of style, unless there are significant performance implications, so I don't feel strongly either way.